### DISCUSSION TITLE
Create an all-in-one initialization script?

### DISCUSSION THREAD

**Discussion by ericbn** [zimfw/zimfw#347] [Ideas]
Regarding having an all-in-one initialization file [suggested before](https://github.com/zimfw/zimfw/pull/323#issuecomment-553930310), solving it for scripts that look at `${0}` is straightforward as can be seen in the solutions below. The problem is when there's a `%N` or `%x` in the script. Even with a complete Zsh parser, I don't think it will be obvious if these are expansions are being used to get the script path, or if they are being used in prompts, or if for any other "creative" use.

NOTE 1: `%N` to get the script path is used as a workaround for when FUNCTION_ARGZERO is unset. This option is only set by default in the csh and zsh emulations, which blows my mind on why someone would run Zsh with a different emulation and expect Zsh plugins to still work. The general way to resolve `%N` into a variable is with `${(%):-%N}`, as posted in the Zsh Mailing List a while ago (could not find the reference now). Also, `${(%):-%x}` can be seen. And maybe there can be other variations, which again make the parsing of the script hard (if not impossible).

NOTE 2: An example of a "creative" usage of `${(%):-%N}`: https://www.zsh.org/mla/users/2019/msg00651.html

Just for the sake of giving names, I'm naming 3 different solutions:

* conservative, the one currently being used: always sources the scripts.

* moderate: still sources scripts that have `%N` or `%x` on them. This can leave out scripts that can potentially be concatenated into the all-in-one script. Maybe concatenated scripts can be broken in ways we still don't predict. Diff from the conservative solution is:
   ```
   72c72,81
   <     print -Rn ${(F):-source ${^_zscripts:P}}
   ---
   >     local zscript
   >     for zscript in ${_zscripts:P}; do
   >       if command grep -q '%[Nx]' ${zscript}; then
   >         print -R "source ${zscript}"
   >       else
   >         print -R "() { local 0=${zscript}"
   >         command cat ${zscript}
   >         print '}'
   >       fi
   >     done
   ```

* aggressive: concatenates all scripts, substituting `${(%):-%N}` or `${(%):-%x}` by `${:-/path/to/script.zsh}`. Possibly concatenated scripts can be broken in ways we still don't predict. Diff from the conservative solution is:
   ```
   72c72,77
   <     print -Rn ${(F):-source ${^_zscripts:P}}
   ---
   >     local zscript
   >     for zscript in ${_zscripts:P}; do
   >       print -R "() { local 0=${zscript}"
   >       command sed "s/\${(%):-%[Nx]}/\${:-${zscript//\//\\/}}/g" ${zscript}
   >       print '}'
   >     done
   ```

I've used the oldest/slowest machine I have here to benchmark the 3 solutions (Intel Core 2 Duo 2.20 GHz, 4 GB RAM, HD, using the first version of WSL which is known to have a poor disk performance). Benchmark done with the default Zim installation (of the develop branch, at the moment of writing this):

Solution | Mean (s) | Min…Max (s)
--- | --- | ---
conservative | 0.1913 ± 0.00364829 | 0.19 … 0.21
moderate | 0.19 ± 0.00141421 | 0.18 … 0.20
aggressive | 0.18543 ± 0.00189871 | 0.183 … 0.200

So, not sure if the difference is worth the risk. And I want to avoid going into the path of giving users the option to set which solution they want, as this will increase the complexity of the project in a way that I consider incompatible with our values.  :- )

@Eriner, want to try to benchmark these in a machine you consider would strongly benefit from the few fopen calls the non-conservative solutions offer?

**Comment by rjcoelho** [comment]
So 6ms diff between conservative and aggressive ? I don't think it's worth it because it increase project maintenance costs and complexity as you pointed out. But it's your call.

**Comment by PatTheMav** [comment]
> So, not sure if the difference is worth the risk. And I want to avoid going into the path of giving users the option to set which solution they want, as this will increase the complexity of the project in a way that I consider incompatible with our values.

Yeah as mentioned in https://github.com/zimfw/zimfw/pull/323#issuecomment-558361941 I kind of expected that, but good to see actual numbers.
